File: python/003. 无重复字符的最长子串.py
"""
@Author: tushushu
@Date: 2018-12-25 10:28:49
"""


class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if s == "":
            return 0
        dic = {}
        ret = 0
        start = 0
        n = len(s)
        for end, c in enumerate(s):
            if c in dic and dic[c] >= start:
                ret = max(ret, end - start)
                start = dic[c] + 1
                if n - start < ret:
                    break
            dic[c] = end
        return max(ret, end - start + 1)


if __name__ == "__main__":
    t = Solution()
    s = " "
    print(t.lengthOfLongestSubstring(s))

File: cpp/003_无重复字符的最长子串.cpp
#include <vector>
#include <unordered_map>
#include <string>
#include <algorithm>

using namespace std;

class Solution
{
  public:
    int lengthOfLongestSubstring(string s)
    {
        // 变量声明
        int ret = 0;
        int start = 0;
        int end = 0;
        int n = s.length();
        char c;
        unordered_map<char, int> mapping;

        // 提前终止
        if (n == 0)
        {
            return ret;
        }

        // 滑动窗口
        while (end < n)
        {
            c = s[end];
            if (mapping.find(c) != mapping.end() && mapping[c] >= start)
            {
                ret = max(ret, end - start);
                start = mapping[c] + 1;
                if (n - start < ret)
                {
                    ret = max(ret, end - start);
                    break;
                }
            }
            mapping[c] = end;
            end++;
        }
        ret = max(ret, end - start);
        return ret;
    }
};